external-crates/move/crates/move-compiler/src/parser/syntax.rs

```diff
@@ -1514,21 +1599,227 @@ fn parse_name_exp(context: &mut Context) -> Result<Exp_, Box<Diagnostic>> {
+//   <MatchArm> = <MatchPat> ("if" <Exp>)? "=>" ("{" <Exp> "}" | <Exp>)
+//
+fn parse_match_arm(context: &mut Context) -> Result<MatchArm, Box<Diagnostic>> {
+    ok_with_loc!(context, {
+        let pattern = parse_match_pattern(context)?;
+        let guard = match context.tokens.peek() {
+            Tok::If => {
+                context.tokens.advance()?;
+                consume_token(context.tokens, Tok::LParen)?;
+                let guard_exp = parse_exp(context)?;
+                consume_token(context.tokens, Tok::RParen)?;
+                Some(Box::new(guard_exp))
+            }
+            _ => None,
+        };
+        consume_token(context.tokens, Tok::EqualGreater)?;
+        let rhs = match context.tokens.peek() {
+            Tok::LBrace => {
+                let block_start_loc = context.tokens.start_loc();
+                context.tokens.advance()?; // consume the LBrace
+                let block_ = Exp_::Block(parse_sequence(context)?);
+                let block_end_loc = context.tokens.previous_end_loc();
+                let exp = spanned(
+                    context.tokens.file_hash(),
+                    block_start_loc,
+                    block_end_loc,
+                    block_,
+                );
+                Box::new(exp)
+            }
+            _ => Box::new(parse_exp(context)?),
+        };
+        MatchArm_ {
+            pattern,
+            guard,
+            rhs,
+        }
+    })
+}
+
+fn parse_match_pattern(context: &mut Context) -> Result<MatchPattern, Box<Diagnostic>> {
+    const INVALID_MUT_ERROR_MSG: &str =
+        "Can't use 'mut' as a modifier outside of variable bindings";
+    const WILDCARD_AT_ERROR_MSG: &str = "Can't use '_' as a binder in an '@' pattern";
+    const INVALID_PAT_ERROR_MSG: &str = "Invalid pattern";
+
+    use MatchPattern_::*;
+
+    fn parse_ctor_pattern(context: &mut Context) -> Result<MatchPattern, Box<Diagnostic>> {
+        match context.tokens.peek() {
+            Tok::LParen => {
+                context.tokens.advance()?;
+                let pat = parse_match_pattern(context);
+                consume_token(context.tokens, Tok::RParen)?;
+                pat
+            }
+            Tok::Identifier if context.tokens.content() == "_" => {
+                context.tokens.advance()?;
+                Ok(sp(
+                    context.tokens.current_token_loc(),
+                    MatchPattern_::Wildcard,
+                ))
+            }
+            Tok::Mut | Tok::Identifier => ok_with_loc!(context, {
+                let mut_ = parse_mut_opt(context)?;
+                let name_access_chain = parse_name_access_chain(context, || "a pattern entry")?;
+                let ty_args = parse_optional_type_args(context)?;
+
+                match mut_ {
+                    Some(loc)
+                        if !matches!(name_access_chain.value, NameAccessChain_::One(_))
+                            || ty_args.is_some() =>
+                    {
```

Comment:
I find this a bit strange. I feel like we should just handl the `Tok::Mut` case separately?

Reply:
This has been revised to match how we handle `let`, with some logic in expansion to handle the last case.